File: swiftAOC/Sources/AocLib/y2020d01.swift
struct Problem202001: Problem {
  var name: String { "2020-01" }
  func solveWith(data: [String]) -> Solution { Solution202001(data: data) }
}

struct Solution202001: Solution {
  let data: [String]

  var part1: String {
    return "\(answer1)"
  }

  var part2: String {
    return "\(answer2)"
  }

  var answer1: Int {
    let nums = data.map { Int($0) ?? 0 }
    for num1 in nums {
      for num2 in nums {
        if num1 + num2 == 2020 { return num1 * num2 }
      }
    }
    return 0
  }

  var answer2: Int {
    let nums = data.map { Int($0) ?? 0 }
    for num1 in nums {
      for num2 in nums {
        let num = num1 + num2
        if num >= 2020 { continue }
        for num3 in nums {
          if num + num3 == 2020 { return num1 * num2 * num3 }
        }
      }
    }
    return 0
  }

}

File: dotnetAOC/y2020d01.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace dotnetAOC
{
    struct Solution202001 : Solution
    {
        public Solution202001(IEnumerable<String> data)
        {
            Data = data;
        }
        private IEnumerable<String> Data { get; set; }

        public String Part1
        {
            get
            {
                return $"{Answer1}";
            }
        }

        public String Part2
        {
            get
            {
                return $"{Answer2}";
            }
        }

        int Answer1
        {
            get
            {
                var diffs = new HashSet<int>();
                var nums = Data.Select(x => int.Parse(x));
                foreach (var num in nums)
                {
                    var diff = 2020 - num;
                    if (diffs.Contains(diff))
                    {
                        return diff * num;
                    }
                    else
                    {
                        diffs.Add(num);
                    }
                }
                return -1;
            }
        }

        int Answer2
        {
            get
            {
                var nums = Data.Select(x => int.Parse(x)).ToArray();
                for (int i = 0; i < nums.Length - 2; i++)
                {
                    for (int j = i + 1; j < nums.Length - 1; j++)
                    {
                        for (int k = j + 1; k < nums.Length; k++)
                        {
                            if (nums[i] + nums[j] + nums[k] == 2020)
                            {
                                return nums[i] * nums[j] * nums[k];
                            }
                        }
                    }
                }
                return -1;
            }
        }
    }
}
